### PR TITLE
Adding changes to support the jwks end point for service provider

### PR DESCRIPTION
**Pull request by SachiniSiriwardene**
### Proposed changes in this pull request

Adding changes to support the jwks end point for service provider
Related-PRs - https://github.com/wso2-support/carbon-identity-framework/pull/434
                        https://github.com/wso2-support/carbon-identity-framework/pull/436

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider-finish-ajaxprocessor.jsp [hunk 1]
```diff
@@ -62,11 +62,20 @@
             } else {
                 String certString = request.getParameter("sp-certificate");
                 String deleteCert = request.getParameter("deletePublicCert");
+
```
Comment (isharak) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider-finish-ajaxprocessor.jsp]: Remove extra new lines

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/java/org/wso2/carbon/identity/application/mgt/ui/ApplicationBean.java [hunk 2]
```diff
@@ -1074,6 +1079,42 @@ public void update(HttpServletRequest request) {
...
+                } else if (!jwks.equals(spProperty.getValue())) {
+                    //if a new value is set for the jwks uri add it to the service provider properties
+                    spPropList.get(spPropList.indexOf(spProperty)).setValue(jwks);
+                    break;
+                }
+
+            }
+        }
+        //when a jwks value is set for the first time add it to spProperties
+        if (!StringUtils.isBlank(jwks) && !jwksExist) {
```
Comment (emswbandara) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/java/org/wso2/carbon/identity/application/mgt/ui/ApplicationBean.java]: Shall we use StringUtils.isNotBlank here?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/java/org/wso2/carbon/identity/application/mgt/ui/ApplicationBean.java [hunk 3]
```diff
@@ -1074,6 +1079,42 @@ public void update(HttpServletRequest request) {
...
+        //if a jwks property already exist for the service provider check whether the stored value is not  equal to the
+        //value received from the request
+        for (ServiceProviderProperty spProperty : spPropList) {
+            if (ApplicationMgtUIUtil.JWKS_URI.equals(spProperty.getName())) {
+                jwksExist = true;
+                if (StringUtils.isBlank(jwks)) {
+                    spPropList.remove(spProperty);
+                    break;
+                } else if (!jwks.equals(spProperty.getValue())) {
+                    //if a new value is set for the jwks uri add it to the service provider properties
```
Comment (emswbandara) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/java/org/wso2/carbon/identity/application/mgt/ui/ApplicationBean.java]: shall we add a debug log here?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp [hunk 4]
```diff
@@ -1205,6 +1224,42 @@
         }
     }
 
+    //
```
Comment (vihanga-liyanage) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp]: Please remove this.

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp [hunk 5]
```diff
@@ -1205,6 +1224,42 @@
...
+    function selectJWKS(certDataNotNull) {
```
Comment (vihanga-liyanage) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp]: Please format the code properly.

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp [hunk 6]
```diff
@@ -1263,7 +1318,27 @@
                                 </div>
                             </td>
                         </tr>
+
+                        <!-----Add radio button to select certificate or jwks end point fro sp--------------->
```
Comment (vihanga-liyanage) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp]: Shall we change this comment to the following?
`<!-- Add a radio button to select the certificate or JWKS endpoint from sp -->`

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp [hunk 7]
```diff
@@ -1263,7 +1318,27 @@
...
+                                    </label>
+                                    <label style="display:block">
+                                    <input type="radio" id="choose_upload_certificate" name="choose_certificate_type"
+                                    <% if (appBean.getServiceProvider().getCertificateContent() != null) { %> checked="checked" <% } %>
+                                    value="choose_upload_certificate" onclick="selectCertificate()" />
+                                Upload SP certificate
+                                    </label>
+                              </td>
+                        </tr>
+                        <!-------------------------->
```
Comment (vihanga-liyanage) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp]: Is this line necessary?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp [hunk 8]
```diff
@@ -1339,6 +1414,18 @@
                                 </div>
                             </td>
                         </tr>
+                        <!-----jwks text box JWKS TEXT BOX------->
```
Comment (vihanga-liyanage) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp]: HTML comments on need `<!--` in the beginning and `-->` in the end.

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp [hunk 9]
```diff
@@ -138,6 +139,24 @@
     boolean isNeedToUpdate = false;
     boolean isAdvanceConsentManagementEnabled = false;
 
+    //adding code to support jwks URI
+    String jwksUri = null;
+    boolean hasJWKSUri = false;
+
+    ServiceProviderProperty[] spProperties = appBean.getServiceProvider().getSpProperties();
+    if (spProperties != null) {
+    for (ServiceProviderProperty spProperty : spProperties) {
```
[the remark this answers is not in the document]
Comment (SachiniSiriwardene) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp]: done

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp (hunk 4 above)
Comment (SachiniSiriwardene) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp]: done

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp (hunk 6 above)
Comment (SachiniSiriwardene) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp]: done

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp (hunk 8 above)
Comment (SachiniSiriwardene) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp]: corrected

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp [hunk 10]
```diff
@@ -138,6 +139,24 @@
     boolean isNeedToUpdate = false;
     boolean isAdvanceConsentManagementEnabled = false;
 
+    //adding code to support jwks URI
+    String jwksUri = null;
+    boolean hasJWKSUri = false;
+
+    ServiceProviderProperty[] spProperties = appBean.getServiceProvider().getSpProperties();
+        if (spProperties != null) {
```
Comment (vihanga-liyanage) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp]: Please reformat.

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp [hunk 11]
```diff
@@ -138,6 +139,24 @@
...
+            for (ServiceProviderProperty spProperty : spProperties) {
+                if (ApplicationMgtUIUtil.JWKS_URI.equals(spProperty.getName())) {
+                    hasJWKSUri = true;
+                    jwksUri = spProperty.getValue();
+                }
+            }
+        }
+
+    if (jwksUri == null) {
+            jwksUri = "";
```
Comment (vihanga-liyanage) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp]: Please reformat.

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider-finish-ajaxprocessor.jsp [hunk 12]
```diff
@@ -66,7 +66,14 @@
                 if (StringUtils.isNotBlank(certString) && !Boolean.parseBoolean(deleteCert)) {
                     CertData certData = IdentityApplicationManagementUtil.getCertData(IdentityUtil.getCertificateString(certString));
                 }
-                appBean.update(request);
+
+                if (request.getParameter("choose_certificate_type").equals("choose_jwks_uri") &&
+                !request.getParameter("jwksUri").equals("") && (!request.getParameter("jwksUri").contains("https"))) {
```
Comment (vihanga-liyanage) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider-finish-ajaxprocessor.jsp]: Shall we stay within 120 characters per line limit?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp [hunk 13]
```diff
@@ -1205,6 +1224,41 @@
...
+            var input = document.createElement('input');
+            input.type = "hidden";
+            input.name = "deletePublicCert";
+            input.id = "deletePublicCert";
+            input.value = "true";
+            document.forms['configure-sp-form'].appendChild(input);
+        }
+    }
+
+    function selectCertificate() {
```
Comment (JKAUSHALYA) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp]: Formatting issue.

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/java/org/wso2/carbon/identity/application/mgt/ui/ApplicationBean.java [hunk 14]
```diff
@@ -1074,6 +1079,44 @@ public void update(HttpServletRequest request) {
         serviceProvider.setApplicationName(request.getParameter("spName"));
         serviceProvider.setDescription(request.getParameter("sp-description"));
         serviceProvider.setCertificateContent(request.getParameter("sp-certificate"));
+        String jwks = request.getParameter(ApplicationMgtUIConstants.JWKS_URI);
+        boolean jwksExist = false;
+        ArrayList<ServiceProviderProperty> spPropList;
+        //adding jwks uri as a sp meta data
```
Comment (JKAUSHALYA) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/java/org/wso2/carbon/identity/application/mgt/ui/ApplicationBean.java]: It is better to have inline comments in the following format. Always start with a capital letter and end with a full stop. 
// \<Comment\>.

Comment (JKAUSHALYA) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/java/org/wso2/carbon/identity/application/mgt/ui/ApplicationBean.java]: Please fix all the places to retain consistancy.

**Pull request merged**